File: fsharp-2017/src/Day02.fs
module Day02

open Util

/// returns difference of max vs. min in xs 
let maxDiff (xs:int[]) = Array.max xs - (Array.min xs)

/// returns first found division of a/b if reminder is 0 in xs
let maxDiv (xs:int[]) = 
  [for a in xs do
    for b in xs do
      if a <> b && a % b = 0 then yield a / b
  ] |> List.head
    
let run (instr:string) =
  let xss = instr |> mapSplitLineAnd '\t' int
  // +1 for compiler suggesting Array.sumBy instead of: let sumOf f xss = xss
  //   |> Array.map (fun xs -> f xs)
  //   |> Array.sum 
  printfn "Day 2-1 answer: %A" <| Array.sumBy maxDiff xss
  printfn "Day 2-1 answer: %A" <| Array.sumBy maxDiv xss

File: js-pipe-fkit_2017/src/day02.js
const { maximum, minimum, fold } = require('fkit')
const { strToInts2D } = require('./helpers')

const maxDiff = xs => maximum(xs) - minimum(xs)

/**
 * @returns {int} 1st result of division of ints in xs where reminder is 0
 */
const maxDiv = xs /*[int]*/ => {
  for (let a of xs)
    for (let b of xs)
      if (a != b && a%b == 0) return a / b
  throw Error(`No a/b == 0 in ${xs}`)
} 

const run = (inStr, log) => {
  const xss = strToInts2D(inStr)
  log.p1(xss |> fold((a, xs) => a + maxDiff(xs), 0)) // 44887
  log.p2(xss |> fold((a, xs) => a + maxDiv(xs), 0))  // 242
}
module.exports = {
  run,
  maxDiff,
  maxDiv,
}